rfcs/01391_async_sdk.md

```diff
@@ -0,0 +1,189 @@
+                reply: format!("HELLO {}!", request.greeting),
+            })?;
+        }).await;
+    }
+}
+```
+
+New functions defined on the existing [`Receiver`](https://project-oak.github.io/oak/sdk/doc/oak/io/struct.Receiver.html) allow asynchronous reads from channels:
+
+```rust
+impl<T: Decodable> Receiver<T> {
+    pub fn receive_async(&self) -> ChannelRead<T> { /* ... */ }
+    pub fn receive_stream(&self) -> ChannelReadStream<T> { /* ... */ }
+}
+```
+
+`ChannelRead` mentioned above is the asynchronous channel read primitive, implementing `Future<Output = Result<T, OakError>`. When `poll`ed, it notifies the executor that it is waiting for a particular channel to return data. The executor is in charge of checking for data on the channel, and waking up the future when the data is ready. 
+
+Since waiting on a channel is the only possible blocking operation (writes do not block in an Oak node), no other primitives are necessary. The `ChannelReadStream` for example is merely a repeated `ChannelRead` implementing `Stream<Item = Result<T, OakError>`.
+
+An `oak::spawn` method is provided that works exactly like [`tokio::spawn`](https://docs.rs/tokio/0.2.22/tokio/fn.spawn.html). When passed a future, a `JoinHandle` is returned that can be used to await the result of the spawned future. If this handle is dropped, the executor keeps polling the future to completion regardless, but there is no way to obtain the result of the computation. 
+
+# Detailed Design
+## Handling invocations
+Instead of a service trait, 
+an enum is generated from the service definition that covers all possible invocations:
+
+```rust
+enum HelloWorld {
+    // Basic request - response
+    SayHello(HelloRequest, OneshotWriter<HelloResponse>),
+    // request - streaming response
+    LotsOfReplies(HelloRequest, MultiWriter<HelloResponse>),
+    // streaming request - single response
+    LotsOfGreetings(
+        RequestStream<HelloRequest>,
+        OneshotWriter<HelloResponse>,
+    ),
+    // streaming request - streaming response
+    BidiHello(
+        RequestStream<HelloRequest>,
+        MultiWriter<HelloResponse>,
+    ),
+}
+```
+
+Each variant is a tuple of (request, response), where the request is either the plain request or a stream of requests, 
+and the `response` is an object that allows writing response data. 
+
+The type of `run_event_loop` shows how the SDK provides a mapping from a gRPC channel to the generated type:
+```
+pub fn run_event_loop<T, F, R>(receiver: Receiver<Invocation>, handler: F)
+where
+    T: TryFrom<oak::grpc::Invocation, OakError>,
+    F: FnOnce(GrpcStream<T>) -> R,
+    R: Future<Output = ()> + 'static,
+```
+
+The `TryFrom` implementation is automatically generated by the service generator.
+
+`MultiWriter`  can be used to send 0 or more messages:
+
+```rust
+impl<T: Message> MultiWriter {
+    // Note: self by reference
+    pub fn send(&self, msg: &T) -> Result<(), OakError> { /* ... */ }
```

Comment:
How does a method invocation get closed out &ndash; implicitly by `drop()`?

Reply:
Maybe we should make it explicit via a `close` method.

Reply:
Yes, my thinking was that `drop()` would close the connection with an OK status.

I forgot to add an explicit `close` method here, will add that too.